### PR TITLE
Add `DatabaseConnectionInfo`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Currently, we have the database information hardcoded. This adds a `DatabaseConnectionInfo` struct, which is used for (a) provide information how to connect to a database and (b) can be used from the command line.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1202510174412962) _(internal)_

## 🔍 What does this change?

- Add the `DatabaseConnectionInfo` struct and use it for connecting
- Change the tests to use `DatabaseConnectionInfo`
- Add `clap` to use `DatabaseConnectionInfo` in the command line

## 🛡 What tests cover this?

The struct is used in the tests

## ❓ How to test this?

Run `cargo run -- --help` in `packages/graph/hash_graph`

## 📹 Demo

image

It's possible to generate completions by running `cargo run -- --generate-completion zsh`:
image

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/lib.rs [hunk 1]
```diff
@@ -38,6 +38,9 @@
     clippy::module_name_repetitions,
     reason = "This encourages importing `as` which breaks IDEs"
 )]
-
+#[allow(
+    clippy::must_use_candidate,
+    reason = "#[must_use] adds boilerplate code, which isn't worth it"
+)]
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/lib.rs:44]: Thoughts? @Alfred-Mountfield @thehabbos007

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/lib.rs:44]: Have you found false positives so far? It sounds like they did a lot of work on reducing them, checking for interior mutability and things

**Review by thehabbos007 — DISMISSED** [review]
Looks great, thanks Tim!

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs [hunk 2]
```diff
@@ -17,6 +17,119 @@ impl fmt::Display for DatastoreError {
...
+            database_type,
+            user,
+            password,
+            host,
+            port,
+            database,
+        }
+    }
+
+    fn url(&self) -> String {
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:89]: Do we perhaps want a docstring saying this is used as a connection URL but shouldn't be used for printing as it includes the password or something. (just a warning)

Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:89]: Do we perhaps want a docstring saying this is used as a connection URL but shouldn't be used for printing as it includes the password or something. (just a warning)

Or do we think the fact that we've implemented `Display` below is enough

File: packages/graph/hash_graph/lib/graph/src/lib.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/lib.rs:44]: And does it make sense for those methods to have `#[must_use]`?

Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/lib.rs:44]: They are getter functions. I'm unsure if we want that. We can still disable the lint later.

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/lib.rs:44]: Why would we want a `get` method without the return value being used? Perhaps it is a good warning

Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/lib.rs]: I added the attributes in 4b966a1baff879fa985e865e597cff2abc2d8d75

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:97]: Addressed in 2ad23eea5fd0e52dd417c3ffc7d812f7a139892f

**Review by Alfred-Mountfield — APPROVED** [review]

**Review by Alfred-Mountfield — APPROVED** [review]

**Pull request merged**